src/qt/listctrl.cpp

```diff
@@ -141,96 +144,868 @@ class wxQtTreeWidget : public wxQtEventSignalHandler< QTreeWidget, wxListCtrl >
+    {
+        return static_cast<int>(m_rows.size());
+    }
+    int columnCount(const QModelIndex& WXUNUSED(parent)) const wxOVERRIDE
+    {
+        return static_cast<int>(m_headers.size());
+    }
+
+    QVariant data(const QModelIndex &index, int role) const wxOVERRIDE
+    {
+        const int row = index.row();
+        const int col = index.column();
+
+        wxCHECK_MSG(row >= 0 && static_cast<size_t>(row) < m_rows.size(),
+            QVariant(),
+            "Invalid row index"
+        );
+
+        const RowItem &rowItem = m_rows[row];
+        const ColumnItem &columnItem = rowItem[col];
+
+        const bool isSelected = IsSelected(index);
+
+        switch ( role )
+        {
+            case Qt::DisplayRole:
+            {
+                return QVariant::fromValue(columnItem.m_label);
+            }
+            case Qt::EditRole:
+            {
+                wxListItem listItem;
+                listItem.SetId(row);
+                listItem.SetColumn(col);
+
+                wxListEvent event(wxEVT_LIST_BEGIN_LABEL_EDIT, m_listCtrl->GetId());
+                event.SetEventObject(m_listCtrl);
+                event.SetItem(listItem); 
+                if ( m_listCtrl->HandleWindowEvent(event) && !event.IsAllowed() )
+                    return QVariant();
+
+                return QVariant::fromValue(columnItem.m_label); 
+            }
+            case Qt::DecorationRole:
+            {
+                wxImageList *imageList = GetImageList();
+                if ( imageList == NULL )
+                    return QVariant();
+
+                int imageIndex = -1;
+
+                if ( columnItem.m_selectedImage != -1 )
+                {
+                    if ( isSelected )
+                        imageIndex = columnItem.m_selectedImage;
+                }
+
+                if ( imageIndex == -1 )
+                    imageIndex = columnItem.m_image;
+
+                if ( imageIndex == -1 )
+                    return QVariant();
+
+                wxBitmap image = imageList->GetBitmap(imageIndex);
+                wxCHECK_MSG(image.IsOk(), QVariant(), "Invalid image");
+                return QVariant::fromValue(*image.GetHandle());
+            }
+            case Qt::FontRole:
+                return QVariant::fromValue(columnItem.m_font);
+
+            case Qt::BackgroundRole:
+                return columnItem.m_backgroundColour.isValid() && !isSelected
+                    ? QVariant::fromValue(columnItem.m_backgroundColour)
+                    : QVariant();
+
+            case Qt::ForegroundRole:
+                return columnItem.m_textColour.isValid() && !isSelected
+                    ? QVariant::fromValue(columnItem.m_textColour)
+                    : QVariant();
+
+            case Qt::TextAlignmentRole:
+                return columnItem.m_align;
+
+             case Qt::CheckStateRole:
+                {
+                  if ( !m_listCtrl->HasCheckBoxes() || col > 0 )
```

Comment:
Maybe it's just me, but the direct check

    if ( col == 0 && m_listCtrl->HasCheckBoxes() )
        return rowItem.m_checked;

seems more clear to me.